src/job/job.ts

```diff
@@ -1,43 +1,237 @@
+      .finally(async () => {
+        this._isRunning = false;
+        await Promise.all([agreementService.end(), networkService.end(), paymentService.end(), marketService.end()]);
+        this.events.emit("ended");
+      });
   }
 
-  async fetchState(): Promise<JobState> {
-    const { state } = await this.fetchFromStorage();
-    return state;
+  private async _runWork({
+    worker,
+    marketService,
+    agreementService,
+    networkService,
+    paymentService,
+    packageDescription,
+    options,
+  }: {
+    worker: Worker<Output>;
+    marketService: MarketService;
+    agreementService: AgreementPoolService;
+    networkService: NetworkService;
+    paymentService: PaymentService;
+    packageDescription: Package;
+    options: RunJobOptions;
+  }) {
+    const allocation = await paymentService.createAllocation();
+
+    await Promise.all([
+      marketService.run(packageDescription, allocation),
+      agreementService.run(),
+      networkService.run(),
+      paymentService.run(),
+    ]);
+    const agreement = await agreementService.getAgreement();
+    // agreement is created, we can stop listening for new proposals
+    await marketService.end();
+
+    paymentService.acceptPayments(agreement);
+
+    this.activity = await Activity.create(agreement.id, this.yagnaApi, options.activity);
+
+    const storageProvider =
+      this.defaultOptions.work?.storageProvider || options.work?.storageProvider || this._getDefaultStorageProvider();
+
+    const workContext = new WorkContext(this.activity, {
+      provider: agreement.provider,
+      storageProvider,
+      networkNode: await networkService.addNode(agreement.provider.id),
+      activityPreparingTimeout:
+        this.defaultOptions.work?.activityPreparingTimeout || options.work?.activityPreparingTimeout,
+      activityStateCheckingInterval:
+        this.defaultOptions.work?.activityStateCheckingInterval || options.work?.activityStateCheckingInterval,
+    });
+
+    this.events.emit("started");
+    await workContext.before();
+    return worker(workContext);
   }
-  async fetchResults(): Promise<Output | undefined> {
-    const { results } = await this.fetchFromStorage();
-    return results;
+  private _getDefaultStorageProvider(): StorageProvider {
+    if (runtimeContextChecker.isNode) {
+      return new GftpStorageProvider();
+    }
+    if (runtimeContextChecker.isBrowser) {
+      return new WebSocketBrowserStorageProvider(this.yagnaApi, {});
+    }
+    return new NullStorageProvider();
   }
-  async fetchError(): Promise<Error | undefined> {
-    const { error } = await this.fetchFromStorage();
-    return error ? new Error(error) : undefined;
+
+  /**
+   * Cancel the job. This method will stop the activity and wait for it to finish.
+   * Throws an error if the job is not running.
+   */
+  async cancel() {
+    if (!this.isRunning || !this.activity) {
+      throw new Error(`Job ${this.id} is not running`);
+    }
+    await this.activity.stop();
```

Comment:
Here we should also release the agreement: `await agreementPoolService.releaseAgreement(...)`